openfe/setup/atommapping.py

```diff
@@ -19,14 +24,99 @@ class AtomMapping:
+
+    def _match_elements(self, idx: int):
+        elem_mol1 = self.mol1.GetAtomWithIdx(idx).GetAtomicNum()
+        elem_mol2 = self.mol2.GetAtomWithIdx(self.mol1_to_mol2[idx]).GetAtomicNum()
+        if elem_mol1 == elem_mol2:
+            return True
+        else:
+            return False
+
+    def _get_unique_bonds_and_atoms(self, mapping: AtomMap):
+
+        uniques = {
+            "atoms": [],  # atoms which fully don't exist in mol2
+            "elements": [],  # atoms which exist but change elements in mol2
+            "bonds": [],  # bonds involving either unique atoms or elements
+        }
+
+        for at in self.mol1.GetAtoms():
+            idx = at.GetIdx()
+            if idx not in self.mol1_to_mol2.keys():
+                uniques["atoms"].append(idx)
+            elif not self._match_elements(idx):
+                uniques["elements"].append(idx)
+
+        for bond in self.mol1.GetBonds():
+            bond_at_idxs = [bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()]
+            for at in chain(uniques["atoms"], uniques["elements"]):
+                if at in bond_at_idxs:
+                    bond_idx = bond.GetIdx()
+                    # avoid double inclusion
+                    if bond_idx not in uniques["bonds"]:
+                        uniques["bonds"].append(bond_idx)
+
+        return uniques
+
+    def _draw_mapping(self):
+        mol1_uniques = self._get_unique_bonds_and_atoms(self.mol1_to_mol2)
+
+        # invert map
+        mol2_to_mol1_map = {v: k for k, v in self.mol1_to_mol2.items()}
+
+        mol2_uniques = self._get_unique_bonds_and_atoms(mol2_to_mol1_map)
+
+        atoms_list = [
+            mol1_uniques["atoms"] + mol1_uniques["elements"],
+            mol2_uniques["atoms"] + mol2_uniques["elements"],
+        ]
+
+        # highlight core element changes differently from unique atoms
+        red = (1, 0, 0, 0.5)
+        green = (0, 1, 0, 1)
+
+        at1_colours = {}
+        for at in mol1_uniques["elements"]:
+            at1_colours[at] = green
+
+        at2_colours = {}
+        for at in mol2_uniques["elements"]:
+            at2_colours[at] = green
+
+        atom_colours = [at1_colours, at2_colours]
+
+        bonds_list = [mol1_uniques["bonds"], mol2_uniques["bonds"]]
+
+        d2d = Chem.Draw.rdMolDraw2D.MolDraw2DCairo(600, 300, 300, 300)
+        d2d.drawOptions().useBWAtomPalette()
+        d2d.drawOptions().continousHighlight = False
+        d2d.drawOptions().setHighlightColour(red)
+        d2d.drawOptions().addAtomIndices = True
+        d2d.DrawMolecules(
+            [self.mol1, self.mol2],
+            highlightAtoms=atoms_list,
+            highlightBonds=bonds_list,
+            highlightAtomColors=atom_colours,
+        )
+        d2d.FinishDrawing()
+        return d2d.GetDrawingText()
+
+    def visualize(self):
+        """Visualize atom mapping in a Jupyter Notebook"""
+        from IPython.display import Image
+
+        return Image(self._draw_mapping())
```

Comment:
So as far as I remember the 2DCairo backend should just output a raw stream for a PNG, maybe we can make this an optional return where we either save to file or visualise?

Reply:
Doesn't `_draw_mapping` already give that stream? I think that it would be better to have a separate method `save_to_file` (returns `None`) instead of having a `visualize` method that gives multiple different output types. `save_to_file` might even parse the filename for the extension, and pick how to draw based on that.

Reply:
> Doesn't _draw_mapping already give that stream?

Yeah that's what I meant.

Separate methods sounds good, I didn't want to pollute the class too much, but if we can move `_draw_mapping` et al (they don't really need `self` so we could just keep `_ipython_display_` and move the rest to `utils`), then adding an additional method here doesn't hurt.

Reply:
Having a  `save_to_file` method sounds good, but I like `visualize` being used to "visualize" inside a notebook, so the CLI can call `save_to_file`